samza-core/src/main/java/org/apache/samza/table/remote/TableAsyncHelper.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.apache.samza.metrics.Timer;
+import org.apache.samza.storage.kv.Entry;
+
+import com.google.common.annotations.VisibleForTesting;
+
+
+/**
+ * Helper class to handle async table IO requests with rate limiting on behalf of the remote tables.
+ *
+ * It provides a set of non-blocking execute() methods to ease the implementation of the different
+ * CRUD operations in table. These async requests are dispatched by a single-threaded executor
```

Comment:
This could be useful for other table types, eg. local, or other tables not remote-table based.

Reply:
Another reason is this helper extracts out the async + rate-limiting logic into a utility class, per table instance, which makes RemoteReadableTable and RemoteReadWriteTable much cleaner and easier to read. 

So I tend to feel we should keep the current structure. But let me know if you do feel strongly about inlining, then I can restructure the code.

Reply:
- It incurs a non-trivial overhead to completely understand this "helper" abstraction since., it involves multiple Function pointers, Futures, Generics and threads. 
- It's only used in one place currently and IMO, we should have a higher bar to include new abstractions to the Samza codebase. If we have a later use-case in mind, we can always refactor and add it back.
- The table is simply passing some of its state to the "helper" and handling errors from the execute methods on the helper. This fudges the roles between these 2 classes
- You can always create utility methods if you prefer on the Table class / static methods under ThrottlerUtils

Reply:
Fair enough, I'll inline it back to RemoteTable then.